lib/spack/spack/util/environment.py

```diff
@@ -400,6 +400,24 @@ def execute(self, env: MutableMapping[str, str]):
         env[self.name] = self.separator.join(directories)
 
 
+class FrozenEnvironment:
+    """Freezes os.environ at the time of instanciation"""
+
+    def __init__(self):
+        self._frozen_environ = os.environ.copy()
+
+    def __enter__(self):
+        self._current_environ = os.environ.copy()
+        os.environ.update(self._frozen_environ)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        os.environ.clear()
+        os.environ.update(self._current_environ)
```

Comment:
`os.environ.update()` adds but doesn't clear out what's already in the environment, so this doesn't really swap between the frozen env and the pre-build env -- it adds frozen environment to it.

I think it's clearer to do this with a context manager function like:

```python
@contextlib.contextmanager
def replace_environment(env: Dict) -> None:
    saved_env = os.environ.copy()
    os.environ.clear()
    os.environ.update(env)

    try:
        yield
    finally:
        os.environ.clear()
        os.environ.update(saved_env)
```

which probably should go in `spack.util.environment`.